lib/yamatanooroti/vterm.rb

```diff
@@ -42,76 +40,121 @@ def write(str)
       end
     end
     @pty_input.write(str_to_write)
-    sync
+    # Write str (e.g. `exit`) to pty_input might terminate the process.
+    try_sync
   end
 
   def close
-    sync
-    @pty_input.close
-    sync
-    Process.kill('KILL', @pid)
-    Process.waitpid(@pid)
+    begin
+      sync
+      @pty_input.close
+      sync
+    rescue IOError, Errno::EIO
+    end
+    begin
+      Process.kill('KILL', @pid)
+      Process.waitpid(@pid)
+    rescue Errno::ESRCH
+    end
   end
 
-  private def sync
-    startup_message = +'' if @startup_message
+  private def wait_startup_message
+    wait_until = Time.now + @timeout
+    chunks = +''
     loop do
-      sleep @wait
-      chunk = @pty_output.read_nonblock(1024)
-      if @startup_message
-        startup_message << chunk
-        if @startup_message.(startup_message)
-          @startup_message = nil
-          chunk = startup_message
-        else
-          redo
-        end
+      wait = wait_until - Time.now
+      if wait.negative? || !@pty_output.wait_readable(wait)
+        raise "Startup message didn't arrive within timeout: #{chunks.inspect}"
       end
-      @vterm.write(chunk)
-      chunk = @vterm.read
-      @pty_input.write(chunk)
-    rescue Errno::EAGAIN, Errno::EWOULDBLOCK
-      retry if @startup_message
-      break
-    rescue Errno::EIO # EOF
-      retry if @startup_message
-      break
-    rescue IO::EAGAINWaitReadable # emtpy buffer
-      retry if @startup_message
-      break
+
+      chunk = @pty_output.read_nonblock(65536)
+      vterm_write(chunk)
+      chunks << chunk
+      break if yield chunks
     end
   end
 
+  private def vterm_write(chunk)
+    @vterm.write(chunk)
+    @pty_input.write(@vterm.read)
+    @result = nil
```

Comment:
Is exception handling no longer necessary?

Reply:
### `@vterm.write(chunk)`
It always succeeds because it is not an IO, it won't be closed.

### Errno::EAGAIN, Errno::EWOULDBLOCK, IO::EAGAINWaitReadable

Not needed because there is always wait_readable before read_nonblock

### Errno::EIO

#### `def write(str)`
Failed write (always called from test) should always raise an error and make the test fail.
```ruby
def test_foo
  write("exit!\n") # Process terminates. pty_input will be closed.
  write("1+2\n") # This should raise error
end
```

#### `@pty_input.write(@vterm.read)`
There is a chance that error might raise. I'll add it.
(It rarely happens. I can't reproduce raising Errno::EIO without inserting `sleep 0.0004` before calling vterm_write)